src/diffusers/models/unet_2d_blocks.py

```diff
@@ -1384,6 +1456,199 @@ def forward(
+                    dropout_after_conv=True,
+                    group_size=resnet_group_size,
+                    conv_shortcut_bias=False,
+                )
+            )
+            attentions.append(
+                BasicTransformerBlock(
+                    out_channels,
+                    out_channels // attn_num_head_channels,
+                    attn_num_head_channels,
+                    dropout=None,
+                    cross_attention_dim=cross_attention_dim,
+                    temb_channels=temb_channels,
+                    attention_bias=True,
+                    attn1_type=attn1_type,
+                    attn2_type=attn2_type,
+                    with_ff=False,
+                    use_conv_proj=True,
+                    cross_attention_norm=True,
+                    attention_dropout=dropout,
+                    group_size=resnet_group_size,
+                    norm_type="ada_group",
+                )
+            )
+
+        self.resnets = nn.ModuleList(resnets)
+        self.attentions = nn.ModuleList(attentions)
+
+        if add_downsample:
+            self.downsamplers = nn.ModuleList([KDownsample2d()])
+        else:
+            self.downsamplers = None
+
+        self.gradient_checkpointing = False
+
+    def forward(
+        self, hidden_states, temb=None, encoder_hidden_states=None, attention_mask=None, cross_attention_kwargs=None
+    ):
+        output_states = ()
+
+        for resnet, attn in zip(self.resnets, self.attentions):
+            if self.training and self.gradient_checkpointing:
+
+                def create_custom_forward(module, return_dict=None):
+                    def custom_forward(*inputs):
+                        if return_dict is not None:
+                            return module(*inputs, return_dict=return_dict)
+                        else:
+                            return module(*inputs)
+
+                    return custom_forward
+
+                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
+                hidden_states = torch.utils.checkpoint.checkpoint(
+                    create_custom_forward(attn, return_dict=False),
+                    hidden_states,
+                    encoder_hidden_states,
+                    attention_mask,
+                    cross_attention_kwargs,
+                )
+            else:
+                hidden_states = resnet(hidden_states, temb)
+                hidden_states = attn(
+                    hidden_states,
+                    encoder_hidden_states=encoder_hidden_states,
+                    emb=temb,
+                    attention_mask=attention_mask,
+                    cross_attention_kwargs=cross_attention_kwargs,
+                )
+
+            if self.downsamplers is None:
+                output_states += (None,)
+            else:
+                output_states += (hidden_states,)
+
+        if self.downsamplers is not None:
```

Comment:
so in k-diffusion it calls the downsampler first - should we move it back again? (it is inserted in `self[0]` )
https://github.com/crowsonkb/k-diffusion/blob/master/k_diffusion/models/image_v1.py#L46

Reply:
As tried to explain offline, it's more important that the results match 1-to-1 instead of exactly following the original architecture. 

Note that the following structures are identical:
1.
```
[resnet, attn] (block1) -> [ downsample -> resnet, attn ] (block2) -> ... -> [downsample -> resnet, attn ] (block4)
```

2.
```
[resnet, attn -> downsample] (block1) -> [ resnet, attn -> downsample ] (block2) -> ... -> [resnet, attn ] (block4)
```

Note that here we want the variation 2) because this way we don't need to do hacky changes in the `UNet2DConditionModel`